srml/staking/src/phragmen.rs

```diff
@@ -16,208 +16,342 @@
 			}
 		} else {
 			// if we have less than minimum, use the previous validator set.
-			elected_candidates = original_candidates;
+			return None
 		}
 	}
+	Some(elected_candidates)
+}
+
+pub fn equalise<T: Trait + 'static>(
+	nominator: &mut Nominator<T::AccountId, BalanceOf<T>>,
+	elected_candidates: &mut Vec<Candidate<T::AccountId, BalanceOf<T>>>,
+	tolerance: BalanceOf<T>
+) -> BalanceOf<T> {
 
-	elected_candidates
-}
\ No newline at end of file
+	let mut elected_edges = nominator.edges
+		.iter_mut()
+		.filter(|e| e.elected)
+		.collect::<Vec<&mut Edge<T::AccountId, BalanceOf<T>>>>();
+	if elected_edges.len() == 0 { return <BalanceOf<T>>::zero(); }
+	let stake_used = elected_edges
+		.iter()
+		.fold(<BalanceOf<T>>::zero(), |s, e| s + e.backing_stake);
+	let backed_stakes = elected_edges
+		.iter()
+		.map(|e| elected_candidates[e.elected_idx].backing_stake)
+		.collect::<Vec<BalanceOf<T>>>();
+	let backing_backed_stake = elected_edges
+		.iter()
+		.filter(|e| e.backing_stake > <BalanceOf<T>>::zero())
+		.map(|e| elected_candidates[e.elected_idx].backing_stake)
+		.collect::<Vec<BalanceOf<T>>>();
+
+	let mut difference;
+	if backing_backed_stake.len() > 0 {
+		let max_stake = backing_backed_stake
+			.iter()
+			.max()
+			.expect("vector with positive length will have a max; qed")
+			.to_owned();
+		let min_stake = backed_stakes
+			.iter()
+			.min()
+			.expect("vector with positive length will have a max; qed")
+			.to_owned();
+		difference = max_stake - min_stake;
+		difference += nominator.budget - stake_used;
+		if difference < tolerance {
+			return difference;
+		}
+	} else {
+		difference = nominator.budget;
+	}
+
+	// Undo updates to exposure
+	elected_edges.iter_mut().for_each(|e| {
+		// NOTE: no assertions in the runtime, but this should nonetheless be indicative.
+		//assert_eq!(elected_candidates[e.elected_idx].who, e.who);
+		elected_candidates[e.elected_idx].backing_stake -= e.backing_stake;
+		elected_candidates[e.elected_idx].exposure.total -= e.backing_stake;
+		e.backing_stake = <BalanceOf<T>>::zero();
+	});
+
+	elected_edges.sort_unstable_by_key(|e| elected_candidates[e.elected_idx].backing_stake);
+
+	let mut cumulative_stake = <BalanceOf<T>>::zero();
+	let mut last_index = elected_edges.len() - 1;
+	let budget = nominator.budget;
+	elected_edges.iter_mut().enumerate().for_each(|(idx, e)| {
+		let stake = elected_candidates[e.elected_idx].backing_stake;
+
+		let stake_mul = stake.checked_mul(&<BalanceOf<T>>::sa(idx as u64)).unwrap_or(<BalanceOf<T>>::max_value());
+		let stake_sub = stake_mul.checked_sub(&cumulative_stake).unwrap_or_default();
+		if stake_sub > budget {
+			last_index = idx.clone().checked_sub(1).unwrap_or(0);
+			return
+		}
+		cumulative_stake += stake;
+	});
+
+	let last_stake = elected_candidates[elected_edges[last_index].elected_idx].backing_stake;
+	let split_ways = last_index + 1;
+	let excess = nominator.budget + cumulative_stake - last_stake * <BalanceOf<T>>::sa(split_ways as u64);
```

Comment:
If this gets slipped into merge without discussing this: We definitely should open an issue to at least keep an eye on the arithmetic here for over/under-flow and how to react to it.